File: benchmarks/bench_json.js
#!/usr/bin/env node
// Node counterpart of bench_json.strada (built-in JSON — note this is
// native C++, vs Strada's pure-Strada module; see BASELINE.md note).
function buildDoc(users) {
    const list = [];
    for (let i = 0; i < users; i++) {
        list.push({
            id: i, name: "user_" + i, email: `user${i}@example.com`,
            active: (i % 3) === 0 ? 1 : 0, score: i * 1.5,
            tags: ["alpha", "beta", "tag" + (i % 50)],
            profile: { city: "city" + (i % 100), zip: 10000 + (i % 90000), langs: ["en", "de"] },
        });
    }
    return { count: users, users: list };
}

const doc = buildDoc(2000);
let t0 = Date.now();
let encLen = 0, json = "";
for (let r = 0; r < 20; r++) { json = JSON.stringify(doc); encLen += json.length; }
let t1 = Date.now();
console.log("encode:", encLen, (t1 - t0) / 1000);

let decUsers = 0;
for (let r = 0; r < 20; r++) { decUsers += JSON.parse(json).count; }
let t2 = Date.now();
console.log("decode:", decUsers, (t2 - t1) / 1000);

let rt = 0;
for (let i = 0; i < 20000; i++) {
    const small = JSON.stringify({ op: "get", id: i, args: [1, 2, 3] });
    rt += JSON.parse(small).id % 2;
}
let t3 = Date.now();
console.log("roundtrip:", rt, (t3 - t2) / 1000);
console.log("total:", (t3 - t0) / 1000);

File: benchmarks/bench_json.go
// Go counterpart of bench_json.strada (stdlib encoding/json — native, vs
// Strada's pure-Strada module; see BASELINE.md note).
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

type Profile struct {
	City  string   `json:"city"`
	Zip   int      `json:"zip"`
	Langs []string `json:"langs"`
}

type User struct {
	ID      int      `json:"id"`
	Name    string   `json:"name"`
	Email   string   `json:"email"`
	Active  int      `json:"active"`
	Score   float64  `json:"score"`
	Tags    []string `json:"tags"`
	Profile Profile  `json:"profile"`
}

type Doc struct {
	Count int    `json:"count"`
	Users []User `json:"users"`
}

func buildDoc(users int) Doc {
	list := make([]User, 0, users)
	for i := 0; i < users; i++ {
		active := 0
		if i%3 == 0 {
			active = 1
		}
		list = append(list, User{
			ID: i, Name: "user_" + strconv.Itoa(i),
			Email:  fmt.Sprintf("user%d@example.com", i),
			Active: active, Score: float64(i) * 1.5,
			Tags:    []string{"alpha", "beta", "tag" + strconv.Itoa(i%50)},
			Profile: Profile{City: "city" + strconv.Itoa(i%100), Zip: 10000 + (i % 90000), Langs: []string{"en", "de"}},
		})
	}
	return Doc{Count: users, Users: list}
}

type Small struct {
	Op   string `json:"op"`
	ID   int    `json:"id"`
	Args []int  `json:"args"`
}

func main() {
	doc := buildDoc(2000)
	t0 := time.Now()
	encLen := 0
	var jsonBytes []byte
	for r := 0; r < 20; r++ {
		jsonBytes, _ = json.Marshal(doc)
		encLen += len(jsonBytes)
	}
	t1 := time.Now()
	fmt.Println("encode:", encLen, t1.Sub(t0).Seconds())

	decUsers := 0
	for r := 0; r < 20; r++ {
		var d Doc
		json.Unmarshal(jsonBytes, &d)
		decUsers += d.Count
	}
	t2 := time.Now()
	fmt.Println("decode:", decUsers, t2.Sub(t1).Seconds())

	rt := 0
	for i := 0; i < 20000; i++ {
		small, _ := json.Marshal(Small{Op: "get", ID: i, Args: []int{1, 2, 3}})
		var s Small
		json.Unmarshal(small, &s)
		rt += s.ID % 2
	}
	t3 := time.Now()
	fmt.Println("roundtrip:", rt, t3.Sub(t2).Seconds())
	fmt.Println("total:", t3.Sub(t0).Seconds())
}
